New NumberControl component

m-e-h opened the pull request:
## Description
Adds a `<NumberControl>` component for use in other components and blocks.

Would reduce duplicate code and create a consistent style across existing components that are creating their own number inputs. Currently, this would be [font-size-picker](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/font-size-picker), [focal-point-picker](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/focal-point-picker), and [range-control](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/range-control) that I know of. Probably others.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

Inspired by https://github.com/WordPress/gutenberg/pull/13709 and https://github.com/WordPress/gutenberg/pull/11555

m-e-h committed:
wider range number controls to account for OS specific input spinners.

m-e-h committed:
Revert "wider range number controls to account for OS specific input spinners."

This reverts commit d26c6595dd46ae101cde8c3065a9e438db3b8c25.

m-e-h committed:
Merge remote-tracking branch 'upstream/master'

m-e-h committed:
Merge remote-tracking branch 'upstream/master'

m-e-h committed:
Merge remote-tracking branch 'upstream/master'

m-e-h committed:
Merge remote-tracking branch 'upstream/master'

m-e-h committed:
New Component: NumberControl

m-e-h committed:
Use NumberControl in Focal Point Picker

m-e-h committed:
Use NumberControl in Font Size Picker

m-e-h committed:
Use NumberControl in RangeControl

m-e-h commented:
Not sure if I should update existing components in the same PR or if they should be done individually after the NumberControl component is merged.
I've added them here to make it easier to test but they can be removed.

jorgefilipecosta commented on packages/components/src/focal-point-picker/index.js:
```diff
@@ -210,27 +211,31 @@ export class FocalPointPicker extends Component {
-							value={ this.fractionToPercentage( percentages.x ) }
-						/>
+
+					<NumberControl
```
It looks like here we have a NumberControl inside a BaseControl. NumberControl renders its own BaseControl component so I think the existing BaseControl can also be removed.

m-e-h replied:
Thanks for looking @jorgefilipecosta! I'm open to any tips or ideas here. I'm pretty new to the components code. 

Yeah, we could do away with the BaseControl around each input here with the focal-point-picker. Will have to restyle though to handle the `<span>%</span>` appended to each.

Might even be worth adding a `numericUnit` prop in the NumberControl to handle it.

jorgefilipecosta commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -88,12 +89,11 @@ function FontSizePicker( {
-						className="components-range-control__number"
-						type="number"
-						onChange={ onChangeValue }
+					<NumberControl
```
With this change the number input field is rendered inside a base control div, that creates an unexpected small design change:
<img width="561" alt="screenshot 2019-02-13 at 18 27 03" src="https://user-images.githubusercontent.com/11271197/52734493-4ab1c700-2fbd-11e9-9ef5-27e6579f3b1b.png">

The input field now is not in the middle of the reset and the size picker because of some padding.

m-e-h replied:
Must've missed that while consolidating the different components CSS. That's actually a margin on the input itself. Just fixed.

jorgefilipecosta reviewed (commented):
Thank you for your contribution @m-e-h, I think it may make sense to expose a very simple component that just renders a number input field.
Using it as the basis for other components may make the work of creating reactive native components simpler as it reduces direct dom usage.
But I would like more thoughts about creating a new component cc: @youknowriad

m-e-h committed:
lint

m-e-h committed:
margin-left only for range component

chrisvanpatten commented on packages/components/src/number-control/style.scss:
```diff
@@ -0,0 +1,18 @@
+	display: inline-block;
+	font-weight: 500;
+	height: 30px;
+	width: 54px;
```
Love the idea of a dedicated NumberControl but nervous about defining the width here. This makes it tougher to use as a full-width control. Maybe there could be a modifier class (`.components-number-control__input--narrow` or something like that) which applies the width constraint?

m-e-h committed:
update range test

m-e-h committed:
fix typo

youknowriad commented:
Noting that the `TextComponent` supports the `type` prop which means this new component is somehow a duplicate?

m-e-h commented:
Whoa. I hadn't even thought of that @youknowriad.

I think there are enough blocks ignoring that and using their own number input that it warrants it's own component. Better to repeat the TextControl once as a new component than to repeat similar code each time a number input is used.

Wondering though if a NumberControl component should just import the `TextControl` and change the `type`? Then it's not really a "duplicate", it would just be a more obvious choice for devs that need it.

chrisvanpatten commented:
I like the idea of a dedicated `NumberControl` because it could allow for additional number functionality in the future (perhaps including custom buttons to increment/decrement, or basic validation, or something along those lines), but I agree; it should pull from the `TextControl` as a base at least until other UI needs become visible.

m-e-h commented:
One challenge I ran into when trying to use the `TextControl` for this, is that it's defining `onChange` in the return rather than in the defautProps. 
This makes it readonly for things like the Focal Point Picker that define their own `onChange={ this.horizontalPositionChanged }`.

It's likely that I'm missing something. Is there a different way I should be doing it?

youknowriad commented:
> One challenge I ran into when trying to use the `TextControl` for this, is that it's defining `onChange` in the return rather than in the defautProps.

That doesn't change anything for me. If the value is not changing, it simply means the `value` prop is not updated properly for me.

afercia commented:
A generic component shouldn't assume an `aria-label` is always desired. Input fields need to be labelled, In order of preference:
- a _visible_ label element associated with for/id attributes: in the BaseControl component this is controlled by the presence of both `label && id` props
- if there's no id, then the label value is rendered as plain text and the input is supposed to be labelled with an 'aria-label` 

Never do both together.

Some controls in the codebase are misleading, for example the `FocalPointPicker` renders two input fields that are completely unlabelled. The visible text is just text. There's no `<label>` element or `aria-label`... going to create an issue.

m-e-h committed:
Merge remote-tracking branch 'upstream/master'

m-e-h committed:
Merge remote-tracking branch 'upstream/master'

m-e-h committed:
Merge branch 'master' into add/component-number-control

m-e-h committed:
Make id and area-label optional

m-e-h committed:
use NumberControl in the spacer block

m-e-h commented:
I updated this to better handle id and aria-label per @afercia 

I'm sticking with wrapping this in it's own `BaseControl`. Pulling in the `TextControl` wasn't matching up the `for=`/`id` attributes and I struggled to find an elegant way to get at the `BaseControl` props.

I also added an optional `numberUnit` prop so a unit of measurement could be added, like the `%` on the focal-point component.

I also updated the spacer block to see how the `NumberControl` worked there.

jorgefilipecosta commented on packages/components/src/number-control/index.js:
```diff
@@ -0,0 +1,53 @@
+				{ ...props }
+			/>
+
+			{ numberUnit && (
```
Maybe we can update the Number control, to pass children to base control. It would allow us to remove the need for a numbeUnit prop making the component simpler. Current usages would only need to pass span prop with the unit like they do now.

jorgefilipecosta commented on packages/components/src/range-control/index.js:
```diff
@@ -89,10 +89,9 @@ function RangeControl( {
 				{ ...props } />
 			{ afterIcon && <Dashicon icon={ afterIcon } /> }
-			<input
+			<NumberControl
```
Range control is a complex control in itself with validation logic that was recently added. The component already uses a BaseControl so using a NumberControl here would result in a BaseControl nested inside another BaseControl. I would prefer if we did not use NumberControl inside RangeControl at least for now.

jorgefilipecosta commented on packages/components/src/number-control/index.js:
```diff
@@ -0,0 +1,53 @@
+	numberUnit,
+	...props
+} ) {
+	const onChangeValue = ( event ) => onChange( Number( event.target.value ) );
```
Maybe we can change this line to:
```
	const onChangeValue = ( event ) => onChange(
		event.target.value === '' ? undefined : Number( event.target.value )
	);
```

The current version causes a bug -- it becomes impossible to have the field empty. It is possible to reproduce the bug testing the spacer.
The bug happens because Number('') === 0;

jorgefilipecosta reviewed (commented):
Hi @m-e-h,
Thank you for iterating on this PR.
I did a set of tests, and this PR seems almost ready, I just left some minor suggestions.

Besides the suggestions, I think this PR needs a "New Features" changelog entry in https://github.com/WordPress/gutenberg/blob/master/packages/components/CHANGELOG.md#720-unreleased,  and a rebase.

ZebulanStanphill commented:
What is the status of this PR?

m-e-h commented:
This can probably be closed. The goal here was mostly to clean up the UI style variations across the different number inputs. 
There are probably simpler ways to do that.
